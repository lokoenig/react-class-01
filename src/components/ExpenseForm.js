import React from "react";
import DatePicker from "react-datepicker";
import "react-datepicker/dist/react-datepicker.css";
export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        let pe = props.expense;
       
        this.state = {
            description: pe ? pe.description : '',
            note: pe ? pe.note : '',
            amount: pe ? (pe.amount/100).toString() : '0',
            created: pe ? pe.created : (new Date()),
            errorMsg: '',
            buttonText: props.buttonText ? props.buttonText : 'Add Expense'
        };
    };

    onChangeDescription = (e) => {
        const newDiscription = e.target.value;
        this.setState(() => ({
            description: newDiscription
        }));
    };

    onChangeNote = (e) => {
        const note = e.target.value;
        this.setState(() => ({
            note
        }));
    };

    onChangeAmount = (e) => {
        const v = e.target.value;
        if (v.match(/^\d*\.?\d{0,2}$/)) {
            this.setState(() => ({
                amount: v
            }));
        }
    };

    onChangeDate = (newValue) => {
        this.setState(() => ({
            created: newValue
        }));

    };

    onSubmit = (e) => {
        e.preventDefault();
        // So some basic validation:
        if (!this.state.description.length || !this.state.amount.length) {
            this.setState(() => ({
                errorMsg: 'I am the Egg Man'
            }));
            console.log('error');

        } else {
            this.setState(() => ({
                errorMsg: ''
            }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                created: this.state.created.valueOf(),
                note: this.state.note
            });
            console.log('onSubmit');

        }
    }

    render() {
        return (
            <>
                <h1>I be a form!</h1>
                {this.state.errorMsg &&
                    <div className="pre-form-error-banner" key="errMsg1">
                        {this.state.errorMsg}
                    </div>
                }
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        value={this.state.description}
                        onChange={this.onChangeDescription}
                        autoFocus
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onChangeAmount}

                    />
                    <textarea
                        placeholder="Note"
                        value={this.state.note}
                        onChange={this.onChangeNote}
                    >
                    </textarea>

                        <DatePicker
                            selected={this.state.created}
                            onChange={this.onChangeDate}
                        />
                    <button>{this.state.buttonText}</button>
                </form>
            </>
        )
    }
}