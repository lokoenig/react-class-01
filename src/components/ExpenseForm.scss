form {
    --form-gap: 1rem;
    .form-input-group{
        display:flex;
        gap: var(--form-gap);

        .form-input{
            &.field-wide{
                flex-grow: 1;
            }
        }
    }
    label {
        color: blue;
        display: block
    }

    

    input#expense-description, #expense-note {
        width: calc(100% - var(--form-gap));
    }
}