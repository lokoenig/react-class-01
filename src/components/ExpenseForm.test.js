import React from "react";
import { render, screen, waitFor, within } from '@testing-library/react';
import user from '@testing-library/user-event';

import ExpenseTestData from "../fixtures/expenses";

import ExpenseForm from "./ExpenseForm";


describe('ExpenseForm', () => {
    const onSubmit = jest.fn();
 
    test('checks if empty ExpenseForm changed', () => {
        const targetExpense = Math.floor(Math.random() * ExpenseTestData.length);

        const { getByText, asFragment } = render(
            <ExpenseForm />
        );
    
        expect(asFragment()).toMatchSnapshot();
    });


test('submits empty ExpenseForm', ()=>{
    const { getByText, asFragment } = render(
        <ExpenseForm
            onSubmit={onSubmit}
         />
    );
    clickSubmitButton();
    const alertElem = screen.getByRole('alert', {
        name: /expense was not submitted/i
    });
    expect(alertElem).toBeInTheDocument(); // error message
});

    test('type', () => {
        render(<textarea />)

        user.type(screen.getByRole('textbox', {
            name: /expense amount/i
        });
        expect(screen.getByRole('textbox')).toHaveValue('Hello,\nWorld!')
    })

});


function clickSubmitButton() {
    user.click(screen.getByRole('button', { name: /Add Expense/}));
}