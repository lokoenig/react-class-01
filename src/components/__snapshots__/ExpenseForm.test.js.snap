// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseForm checks if ExpenseForm changed 1`] = `
<DocumentFragment>
  <h1>
    I be a form!
  </h1>
  <form
    title="Edit An Expense Entry"
  >
    <input
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      placeholder="Amount"
      type="text"
      value="0"
    />
    <textarea
      placeholder="Note"
    />
    <div
      class="react-datepicker-wrapper"
    >
      <div
        class="react-datepicker__input-container"
      >
        <span
          aria-live="polite"
          class="react-datepicker__aria-live"
          role="alert"
        />
        <input
          class=""
          type="text"
          value="02/09/2023"
        />
      </div>
    </div>
    <button>
      Add Expense
    </button>
  </form>
</DocumentFragment>
`;

exports[`ExpenseForm checks if empty ExpenseForm changed 1`] = `
<DocumentFragment>
  <h1>
    I be a form!
  </h1>
  <form
    title="Edit An Expense Entry"
  >
    <input
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      placeholder="Amount"
      type="text"
      value="0"
    />
    <textarea
      placeholder="Note"
    />
    <div
      class="react-datepicker-wrapper"
    >
      <div
        class="react-datepicker__input-container"
      >
        <span
          aria-live="polite"
          class="react-datepicker__aria-live"
          role="alert"
        />
        <input
          class=""
          type="text"
          value="02/09/2023"
        />
      </div>
    </div>
    <button>
      Add Expense
    </button>
  </form>
</DocumentFragment>
`;
